app/javascript/validate-paired-tubes/components/ValidatePairedTubes.vue

```diff
@@ -0,0 +1,154 @@
+            :api="devourApi"
+            label="Pool XP"
+            :label-cols="3"
+            :fields="devourFields"
+            :includes="devourIncludes"
+            :scan-disabled="allValid"
+            :validators="sourceTubeValidators"
+            labware-type="tube"
+            @change="updateSourceTube($event)"
+          />
+          <lb-labware-scan
+            key="destination"
+            :api="devourApi"
+            label="Pool Norm"
+            :label-cols="3"
+            :fields="devourFields"
+            :includes="devourIncludes"
+            :scan-disabled="allValid"
+            :validators="destinationTubeValidators"
+            labware-type="tube"
+            @change="updateDestinationTube($event)"
+          />
+        </div>
+        <p
+          v-show="allValid"
+          class="tv-instructions"
+        >
+          The scanned tubes match. Transfer volumes are now shown. Refresh the page to scan another pair.
+        </p>
+      </b-card>
+    </lb-sidebar>
+  </lb-page>
+</template>
+
+<script>
+import devourApi from 'shared/devourApi'
+import filterProps from './filterProps'
+import LabwareScan from 'shared/components/LabwareScan'
+import LoadingModal from 'shared/components/LoadingModal'
+import resources from 'shared/resources'
+import TransferVolumes from './TransferVolumes'
+import {
+  checkId,
+  checkMolarityResult,
+  checkPurpose,
+  checkState,
+  checkTransferParameters
+} from 'shared/components/tubeScanValidators'
+
+export default {
+  name: 'ValidatePairedTubes',
+  components: {
+    'lb-labware-scan': LabwareScan,
+    'lb-loading-modal': LoadingModal,
+    'lb-transfer-volumes': TransferVolumes
+  },
+  props: {
+    // Sequencescape API V2 URL.
+    sequencescapeApi: { type: String, default: 'http://localhost:3000/api/v2' },
+
+    // Purpose config JSON string of objects keyed by purpose UUIDs.
+    purposeConfigJson: { type: String, required: true }
+  },
+  data () {
+    return {
+      // Devour API object to deserialise assets from sequencescape API.
+      // (See ../../shared/resources.js for details)
+      devourApi: devourApi({ apiUrl: this.sequencescapeApi }, resources),
+
+      // Flag for toggling loading screen
+      loading: false,
+
+      // Message to be shown during loading screen
+      progressMessage: '',
+
+      // The tube scanned as source for the validation.
+      // Also the source of transfer volume inputs.
+      sourceTube: { state: 'empty', labware: null },
+
+      // The tube scanned as source for the validation.
+      // Also the source of transfer volume inputs.
+      destinationTube: { state: 'empty', labware: null }
+    }
+  },
+  computed: {
+    sourceTubeValidators() {
+      return [checkState(['passed']), checkTransferParameters(this.purposeConfigs), checkMolarityResult()]
+    },
+    destinationTubeValidators() {
+      const allowedIdsFromSource = this.sourceTube.labware?.receptacle.downstream_tubes.map(tube => tube.id)
```

Comment:
Ahh, nice, I didn't know operator chaining terminated the evaluation of the rest of the expression. I assumed it followed a similar pattern to ruby where it passed undefined to the right hand side of the expression.

I'm actually having a hard time thinking of a situation where the ruby behaviour would be required.